lib/bolt/plugin.rb

```diff
@@ -277,6 +278,30 @@ def resolve_top_level_references(data)
     # Evaluates a single reference. The value returned may be another
     # reference.
     def resolve_single_reference(reference)
+      if (ttl = cache?(reference))
+        require 'digest/bubblebabble'
+
+        # Luckily we don't need to use a serious hash algorithm
+        reference.delete('cache')
+        id = Digest::SHA2.bubblebabble(reference.sort.to_s)[0..20]
+        FileUtils.mkdir_p(@config.project.cache_file.dirname)
+        FileUtils.touch(@config.project.cache_file)
+        cache = read_cache
+
+        if cache[id] &&
+           Time.now - Time.parse(cache[id]['mtime']) < ttl
+
+          # First remove any cache entries past their ttl
+          # This prevents removing plugins from leaving orphaned cache entries
+          cache.delete_if do |_, entry|
+            Time.now - Time.parse(entry['mtime']) >= entry['ttl']
+          end
+          File.write(@config.project.cache_file, cache.to_json)
+
+          return cache[id]['result']
+        end
+      end
```

Comment:
I think there's enough going on here that it should be separated out into a separate method.